hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
-      Right bs -> pure $ Cardano.AssetName bs
-
-instance ToJSONKey Cardano.AssetName where
-  toJSONKey = toJSONKeyText $ \(Cardano.AssetName bytes) -> decodeUtf8 $ Base16.encode bytes
+-- TODO: Maybe consider using 'renderValue' from cardano-api instead?
+prettyValue :: Value -> Text
+prettyValue value =
+  let Lovelace lovelace = fromMaybe 0 (valueToLovelace value)
+      (ada, decimal) = lovelace `quotRem` 1000000
+      n = length (valueToList value)
+   in unwords $
+        [ show ada <> "." <> padLeft '0' 6 (show decimal)
+        , "₳"
+        ]
+          ++ if n == 0
+            then mempty
+            else ["and", show n, "asset(s)"]
 
-instance FromJSONKey Cardano.AssetName where
-  fromJSONKey = FromJSONKeyTextParser nameFromText
-   where
-    nameFromText t =
-      case Base16.decode (encodeUtf8 t) of
-        Left e -> fail $ "decoding base16: " <> show e
-        Right bytes -> pure $ Cardano.AssetName bytes
-
-genOutput :: Cardano.VKey 'Cardano.Payment StandardCrypto -> Gen TxOut
-genOutput vk =
-  -- NOTE: Scaling a bit the generator to get non-trivial outputs with some
-  -- funds, and not just a few lovelaces.
-  Cardano.TxOut (mkVkAddress vk) <$> scale (* 8) arbitrary
-
----
---- Certificates
----
-
-instance ToJSON (Cardano.DCert StandardCrypto) where
-  toJSON = String . decodeUtf8 . Base16.encode . serialize'
-
-instance FromJSON (Cardano.DCert StandardCrypto) where
-  parseJSON = withText "DCert" $ \t ->
-    case Base16.decode $ encodeUtf8 t of
-      Left err -> fail $ show err
-      Right bs' -> case decodeFull (fromStrict bs') of
-        Left err -> fail $ show err
-        Right v -> pure v
+-- | Pad a text-string to left with the given character until it reaches the given
+-- length.
+--
+-- NOTE: Truncate the string if longer than the given length.
+-- TODO: Move into a separate module.
+padLeft :: Char -> Int -> Text -> Text
+padLeft c n str = T.takeEnd n (T.replicate n (T.singleton c) <> str)
 
 --
--- Utxo
+-- Generators
 --
 
-instance Semigroup (Cardano.UTxO (MaryEra crypto)) where
-  Cardano.UTxO u1 <> Cardano.UTxO u2 = Cardano.UTxO (u1 <> u2)
+genKeyPair :: Gen (VerificationKey PaymentKey, SigningKey PaymentKey)
+genKeyPair = do
+  -- NOTE: not using 'genKeyDSIGN' purposely here, it is not pure and does not
+  -- play well with pure generation from seed.
+  sk <- fromJust . CC.rawDeserialiseSignKeyDSIGN . fromList <$> vectorOf 64 arbitrary
+  let vk = CC.deriveVerKeyDSIGN sk
+  pure (PaymentVerificationKey (Ledger.VKey vk), PaymentSigningKey sk)
+
+-- TODO: Generate non-genesis transactions for better coverage.
+-- TODO: Enable Alonzo-specific features. We started off in the Mary era, and
+-- some of our tests / interfaces aren't fully ready for Alonzo-specific
+-- details. We later changed the ledger's internals to work with Alonzo-era
+-- specific types, and, to make this in incremental steps, this function still
+-- generates Mary transactions, but cast them to Alonzo's.
```

Comment:
:+1:

Reply:
How are the passed `Utxo` utilized? Are all of them spent? or just some arbitrary ones of them?

Reply:
Most of them are spent.